compiler/src/dotty/tools/io/FileWriters.scala

```diff
@@ -25,13 +24,107 @@ import java.util.zip.CRC32
   type InternalName = String
   type NullableFile =  AbstractFile | Null
 
+  inline def ctx(using ReadOnlyContext): ReadOnlyContext = summon[ReadOnlyContext]
+
+  sealed trait DelayedReporting {
+    def hasErrors: Boolean
+    def error(message: Context ?=> Message, position: SourcePosition): Unit
+    def warning(message: Context ?=> Message, position: SourcePosition): Unit
+    def log(message: String): Unit
+
+    def error(message: Context ?=> Message): Unit = error(message, NoSourcePosition)
+    def warning(message: Context ?=> Message): Unit = warning(message, NoSourcePosition)
+  }
+
+  final class EagerDelayedReporting(using captured: Context) extends DelayedReporting:
+    private var _hasErrors = false
+
+    def hasErrors: Boolean = _hasErrors
+
+    def error(message: Context ?=> Message, position: SourcePosition): Unit =
+      report.error(message, position)
+      _hasErrors = true
+
+    def warning(message: Context ?=> Message, position: SourcePosition): Unit =
+      report.warning(message, position)
+
+    def log(message: String): Unit = report.echo(message)
+
+  final class BufferingDelayedReporting extends DelayedReporting {
+    // We optimise access to the buffered reports for the common case - that there are no warning/errors to report
+    // We could use a listBuffer etc - but that would be extra allocation in the common case
+    // Note - all access is externally synchronized, as this allow the reports to be generated in on thread and
+    // consumed in another
+    private var bufferedReports = List.empty[Report]
+    private var _hasErrors = false
+    enum Report(val relay: Context ?=> BackendReporting => Unit):
+      case Error(message: Context => Message, position: SourcePosition) extends Report(ctx ?=> _.error(message(ctx), position))
+      case Warning(message: Context => Message, position: SourcePosition) extends Report(ctx ?=> _.warning(message(ctx), position))
+      case Log(message: String) extends Report(_.log(message))
+
+    def hasErrors: Boolean = synchronized:
+      _hasErrors
+
+    def error(message: Context ?=> Message, position: SourcePosition): Unit = synchronized:
+      bufferedReports ::= Report.Error({case given Context => message}, position)
+      _hasErrors = true
+
+    def warning(message: Context ?=> Message, position: SourcePosition): Unit = synchronized:
+      bufferedReports ::= Report.Warning({case given Context => message}, position)
+
+    def log(message: String): Unit = synchronized:
+      bufferedReports ::= Report.Log(message)
+
+    /** Should only be called from main compiler thread. */
+    def relayReports(toReporting: BackendReporting)(using Context): Unit = synchronized:
+      if bufferedReports.nonEmpty then
+        bufferedReports.reverse.foreach(_.relay(toReporting))
+        bufferedReports = Nil
+  }
+
+  trait ReadSettings:
+    def jarCompressionLevel: Int
+    def debug: Boolean
+
+  trait ReadOnlyContext:
+
+    val settings: ReadSettings
```

Comment:
That blank line is actively confusing. ;)
```suggestion
  trait ReadOnlyContext:
    val settings: ReadSettings
```